base/pmap.jl

```diff
@@ -0,0 +1,102 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+
+"""
+    pgenerate([::WorkerPool], f, c...) -> iterator
+
+Apply `f` to each element of `c` in parallel using available workers and tasks.
+
+For multiple collection arguments, apply f elementwise.
+
+Results are returned in order as they become available.
+
+Note that `f` must be made available to all worker processes; see
+[Code Availability and Loading Packages](:ref:`Code Availability
+and Loading Packages <man-parallel-computing-code-availability>`)
+for details.
+"""
+function pgenerate(p::WorkerPool, f, c)
+    if length(p) == 0
+        return asyncgenerate(f, c)
+    end
+    batches = batchsplit(c, min_batch_count = length(p) * 3)
+    return flatten(asyncgenerate(remote(p, b -> asyncmap(f, b)), batches))
+end
+
+pgenerate(p::WorkerPool, f, c1, c...) = pgenerate(p, a->f(a...), zip(c1, c...))
+
+pgenerate(f, c) = pgenerate(default_worker_pool(), f, c...)
+pgenerate(f, c1, c...) = pgenerate(a->f(a...), zip(c1, c...))
+
+
+
+"""
+    pmap([::WorkerPool], f, c...) -> collection
+
+Transform collection `c` by applying `f` to each element using available
+workers and tasks.
+
+For multiple collection arguments, apply f elementwise.
+
+Note that `f` must be made available to all worker processes; see
+[Code Availability and Loading Packages](:ref:`Code Availability
+and Loading Packages <man-parallel-computing-code-availability>`)
+for details.
+"""
+pmap(p::WorkerPool, f, c...) = collect(pgenerate(p, f, c...))
+#pmap(f, c...) = pmap(default_worker_pool(), f, c...)
+
+function pmap(f, c...; err_retry=nothing, err_stop=nothing, pids=nothing)
+
+    if err_retry != nothing
+        depwarn("`err_retry` is deprecated, use `pmap(retry(f), c...)`.", :pmap)
+        if err_retry == true
+            f = retry(f)
+        end
+    end
+
+    if err_stop != nothing
+        depwarn("`err_stop` is deprecated, use `pmap(@catch(f), c...).", :pmap)
+        if err_stop == false
+            f = @catch(f)
+        end
+    end
+
+    if pids == nothing
+        p = default_worker_pool()
+    else
+        depwarn("`pids` is deprecated, use `pmap(::WorkerPool, f, c...).", :pmap)
+        p = WorkerPool(pids)
+    end
+
+    return pmap(p, f, c...)
+end
+
+
+
+"""
+    batchsplit(c; min_batch_count=1, max_batch_size=100) -> iterator
+
+Split a collection into at least `min_batch_count` batches.
+
+Equivalent to `partition(c, max_batch_size)` when `length(c) >> max_batch_size`.
+"""
+function batchsplit(c; min_batch_count=1, max_batch_size=100)
+
+    # FIXME Use @require per https://github.com/JuliaLang/julia/pull/15495
+    @assert min_batch_count > 0
+    @assert max_batch_size > 1
+
+    # Split collection into batches, then peek at the first few batches...
+    batches = partition(c, max_batch_size)
+    head, tail = head_and_tail(batches, min_batch_count)
+
+    # If there are not enough batches, use a smaller batch size...
+    if length(head) < min_batch_count
```

Comment:
`batches` could be infinite, or it could be expensive or slow to compute the length of `batches`. e.g. If `c` is a stream of database query results coming form a remote server. The point of `head_and_tail` is to collect enough items form `c` to know what batch size to use.